### PR TITLE
Optional flatten

### PR DESCRIPTION
**Pull request by ricellis**
Thanks for your hard work, please ensure all items are complete before opening.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://github.com/cloudant-labs/kafka-connect-cloudant/blob/master/DCO1.1.txt)
- [x] You have added tests for any code changes
- [x] You have updated the [CHANGES.md](https://github.com/cloudant-labs/kafka-connect-cloudant/blob/master/CHANGES.md)
- [x] You have completed the PR template below:

*Note for reviewers:* the changes to `CloudantSourceTask` are much easier to review if excluding the whitespace changes from ba08a2a.

## What

Added configuration option `cloudant.value.schema.struct.flatten` for flattening nested objects and arrays when using `cloudant.value.schema.struct=true`.

## How

* Added option and associated constants and documentation.
* Added new classes for JSON to `Struct` conversion:
    * Created classes for JSON collection to other type conversion.
    * Created overrides for flattening.
    * Created document level converter.
* Enhanced array element detection and added merging for objects in arrays when using structs.
* Fixed ordering requirements for `Struct` and added a note to documentation.
* Enabled flatten option in `CloudantSourceTask`
    * Added option to `CloudantSourceTask`.
    * Called new `DocumentAsSchemaStruct` class.
    * Removed now unused `JsonStruct` class.

* Also fixed whitespace and updated copyright in `CloudantSourceTask`

## Testing

* Added new option to `test.properties` and assertions for `CloudantSourceConnectorTest` - also updated `cloudant.value.schema.struct` to be supplied in source properties only.
* Renamed `JsonStructTest` to `JsonConverterTests` and updated to use new types and added some extra assertions.
*Added new tests to `CloudantSourceTaskTest` for flattening and added some additional assertions.

## Issues

N/A

### CODE REVIEW

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java [hunk 1]
```diff
@@ -29,64 +27,72 @@
...
+                .define(InterfaceConst.TOPIC, Type.LIST,
+                        Importance.HIGH,
+                        ResourceBundleUtil.get(MessageKey.KAFKA_TOPIC_LIST_DOC),
+                        DATABASE_GROUP, 1, Width.LONG,
+                        ResourceBundleUtil.get(MessageKey.KAFKA_TOPIC_LIST_DISP))
+
+                // Whether to generate a struct Schema
+                .define(InterfaceConst.USE_VALUE_SCHEMA_STRUCT, Type.BOOLEAN,
+                        false,
+                        Importance.HIGH,
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java]: Does this need to be `Importance.HIGH` if it's optional?

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java [hunk 2]
```diff
@@ -29,64 +27,72 @@
...
+                        ResourceBundleUtil.get(MessageKey.CLOUDANT_STRUCT_SCHEMA_DOC),
+                        DATABASE_GROUP, 1, Width.NONE,
+                        ResourceBundleUtil.get(MessageKey.CLOUDANT_STRUCT_SCHEMA_DISP))
+
+                // Whether to flatten nested objects in the generated struct
+                .define(InterfaceConst.FLATTEN_VALUE_SCHEMA_STRUCT, Type.BOOLEAN,
+                        false,
+                        Importance.HIGH,
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java]: Same as above

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java [hunk 3]
```diff
@@ -42,210 +40,218 @@
...
+    private AtomicBoolean _running;
+
+    private CloudantSourceTaskConfig config;
+
+    String url = null;
+    String userName = null;
+    String password = null;
+    List<String> topics = null;
+    int taskNumber;
+    boolean generateStructSchema = false;
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: Can the access level for variables on lines 55-60 be set to private?

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java [hunk 4]
```diff
@@ -42,210 +40,218 @@
...
+
+    private static String latestSequenceNumber = null;
+    private static int batch_size = 0;
+
+    private static CloudantClient cantClient = null;
+    private ChangesResult cantChangeResult = null;
+    private static Database cantDB = null;
+    private static String cantDBName = null;
+
+    public int counter = 0;
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: This variable doesn't seem to be used in this class

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java [hunk 5]
```diff
@@ -42,210 +40,218 @@
...
-			}
-		}
-	}
-	
+
+            _running = new AtomicBoolean(false);
+            stop = new AtomicBoolean(false);
+
+            if (latestSequenceNumber == null) {
+                latestSequenceNumber = new String(DEFAULT_CLOUDANT_LAST_SEQ);
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: I believe the new String(...) is redundant and we can just do `latestSequenceNumber = DEFAULT_CLOUDANT_LAST_SEQ`.

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java [hunk 6]
```diff
@@ -42,210 +40,218 @@
...
+        if (cantDB.changes() != null) {
+            cantDB.changes().stop();
+
+            // graceful shutdown
+            // allow the poll() method to flush records first
+            if (_running == null) {
+                return;
+            }
+
+            while (_running.get() == true) {
```
Comment (emlaver) [inline src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: We can remove `== true`  
e.g. `while (_running.get()) {`

**Comment by emlaver** [comment]
FYI comments above for `CloudantSourceTask` are for the legacy code that's outside of this PR.

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java (hunk 1 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java]: Fixed in e44c82e62ea989e9a697167a8f14e44512609c6c

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java (hunk 2 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceConnectorConfig.java]: Fixed in e44c82e62ea989e9a697167a8f14e44512609c6c

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java (hunk 3 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: Fixed in 7c205d5719d12ff03091d5149405206e2ae498d5

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java (hunk 4 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: Fixed in 7c205d5719d12ff03091d5149405206e2ae498d5

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java (hunk 5 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: Fixed in 7c205d5719d12ff03091d5149405206e2ae498d5

File: src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java (hunk 6 above)
Comment (ricellis) [reply src/main/java/com/ibm/cloudant/kafka/connect/CloudantSourceTask.java]: Fixed in 7c205d5719d12ff03091d5149405206e2ae498d5 along with a few other things.

**Comment by ricellis** [comment]
@emlaver I noticed a test failure locally, which I think was due to not all of the docs having the same fields available to assert, so I improved the `CloudantSourceTasksTests` slightly in 7146e9728c697b8ec95a6f1dc1de0015caf54b0c.

**Comment by emlaver** [comment]
Please add details for `cloudant.value.schema.struct.flatten` parameter in Cloudant as source README table.

**Comment by ricellis** [comment]
@emlaver done in aca6081094ecdb4215f7c089178fd17e44984501

**Pull request merged**